lib/screens/v2/disruption_diagram/model.ex

```diff
@@ -56,26 +61,179 @@ defmodule Screens.V2.DisruptionDiagram.Model do
+  end
+
+  # Red Line and Orange Line diagrams never need to be reversed--we just need to fit regions.
+  defp serialize_by_line(_orange_or_red, builder) do
+    builder
+    |> fit_regions()
+    |> Builder.serialize()
+  end
+
+  defp fit_regions(builder) do
+    with :unchanged <- fit_closure_region(builder),
+         :unchanged <- fit_gap_region(builder),
+         :unchanged <- pad_slots(builder) do
+      builder
+    else
+      {:done, builder} -> builder
+    end
+  end
+
+  # The diagram needs to be flipped whenever it's not a GLX-only alert.
+  defp maybe_reverse_gl(builder) do
+    if Builder.glx_only?(builder) do
+      builder
+    else
+      Builder.reverse(builder)
+    end
+  end
+
+  defp fit_closure_region(builder) do
+    current_closure_count = Builder.closure_count(builder)
+
+    with true <- current_closure_count > @max_closure_count,
+         target_closure_slots = 12 - min_non_closure_slots(builder),
+         true <- target_closure_slots < current_closure_count do
+      builder =
+        builder
+        |> Builder.omit_stops(:closure, target_closure_slots)
+        |> minimize_gap()
+
+      {:done, builder}
+    else
+      _ ->
+        :unchanged
+    end
+  end
+
+  defp minimize_gap(builder) do
+    current_gap_count = Builder.gap_count(builder)
+    target_gap_slots = min_gap(builder)
+
+    if target_gap_slots < current_gap_count do
+      Builder.omit_stops(builder, :gap, target_gap_slots)
+    else
+      builder
+    end
+  end
+
+  defp fit_gap_region(builder) do
+    current_gap_count = Builder.gap_count(builder)
+
+    with true <- current_gap_count >= 3,
+         taken_slots = non_gap_slots(builder),
+         baseline = baseline_slots(Builder.closure_count(builder)),
+         target_gap_slots = baseline - taken_slots,
+         true <- target_gap_slots < current_gap_count do
+      builder = Builder.omit_stops(builder, :gap, target_gap_slots)
+
+      {:done, builder}
+    else
+      _ ->
+        :unchanged
+    end
+  end
+
+  defp pad_slots(builder) do
+    current_slot_count = Builder.slot_count(builder)
+
+    if current_slot_count < @minimum_slot_count do
+      {:done, Builder.add_slots(builder, @minimum_slot_count - current_slot_count)}
+    else
+      :unchanged
+    end
+  end
+
+  defp min_non_closure_slots(builder) do
+    Builder.end_count(builder) + Builder.current_location_count(builder) + min_gap(builder)
+  end
+
+  # Number of slots used by all regions except the gap, when it doesn't get minimized
+  defp non_gap_slots(builder) do
+    Builder.end_count(builder) + Builder.closure_count(builder) +
+      Builder.current_location_count(builder)
+  end
+
+  # The minimum possible size of the gap region.
+  defp min_gap(builder) do
+    min(Builder.gap_count(builder), @collapsed_gap_max)
+  end
+
+  for {closure, baseline} <- %{2 => 10, 3 => 10, 4 => 12, 5 => 12, 6 => 14, 7 => 14, 8 => 14} do
+    defp baseline_slots(unquote(closure)), do: unquote(baseline)
```

Comment:
Yeah, it's a very simple use of metaprogramming. It's a way to generate a bunch of repetitive function clauses when the function essentially does a map lookup.

It generates:
```ex
defp baseline_slots(2), do: 10
defp baseline_slots(3), do: 10
defp baseline_slots(4), do: 12
...
```

The end result is functionally about the same as this:
```ex
@closure_to_baseline %{2 => 10, 3 => 10, 4 => 12, 5 => 12, 6 => 14, 7 => 14, 8 => 14}

defp baseline_slots(closure) do
  Map.fetch!(@closure_to_baseline, closure)
```

except it has a little more opportunity to be compile-time optimized, and if we call the function on a value that's not in the mapping, a `FunctionClauseError` is raised instead of a `KeyError`.

Overall... not really necessary, but I find that it feels a little bit more readable than the map-lookup version, once you get used to the `unquote` stuff.

Reply:
ohhh I see. That is pretty neat! I agree it's a bit more readable, as long as you understand the meta-programming basics